readme.txt
lphant Copyright (C)2003 Juanjo < j_u_a_n_j_o@users.sourceforge.net / http://lphant.sourceforge.net >
 
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

-----------------------------------------------------------------------------------


Welcome to lphant, a filesharing client based on the eDonkey2000(C) network.


Visit us at
 http://www.lphant.com
and
 http://sourceforge.net/projects/lphant

The official Forum is also at http://www.lphant.com

Please remember that this is an early version, not all features are implemented yet,
and not everything might work as expected.

You will need Visual Studio 2003 or the .NET Framework 1.1 SDK to compile lphant. 
If you are interested in compile it under MONO or DOTGNU you can work in this task and 
let me know. We will include your changes in the next official release.

You can also compile lphant with SharpDevelop and .NET framework 1.1 SDK.


changelog.txt
v1.01
Juanjo: download 1st and last chunks first to preview sooner
Juanjo: Added CCH (Crumbs corruption handling) lphant uses edonkey crumbs hashes to recover corrupted parts,
		now bigger corrupted block is about 400KB instead of 9,2MB
Juanjo: Start next stopped file when other ends, now lphant selects the highest priority file from the same category,
		if none found, lphant starts the highest priority file.
Juanjo: option to select the max sources limit for each file
Juanjo: Max upload limit to allow unlimited download lowered from 10 to 9 to help spanish 512/128 users. Please set your upload as high as you can while your downloads are not affected.
toertchn: Added multiline linkdrop to client
70n1: speed control from tray menu and configuration of alternative connections speeds in the tray menu

Visual changes:
Juanjo: new lphant icon added [thanks to Psy]
Juanjo: new about box [design by 70n1]
Juanjo: friends list resizes automatically to the window height
Juanjo: lower part in upload form can be resized with a splitter
Juanjo: added some icons to the interface
Juanjo: Do not allow 0 rates in up or down

Fixes:
70n1: fixed, incorrect values for position and size of the windows loading xml values
Juanjo: fixed, Show/Hide details on downloads context menu works again
toertchn: several fixes to run the kernel on mono
Juanjo: fixed, changing the udp port several times while lphant is running causes an exception
Juanjo: fixed, friends list scrollbar is flat again
Juanjo: added splitter to servers form to allow to resize the list and the window log
70n1: fixed rare exception receiving chunk status
Juanjo: fixed, comment and rating values now use string named tags to avoid conflicts with emule
Juanjo: fixed bug in parameter writer
Juanjo: fixed, now it is safe to open download form in forms designer

v1.00
Juanjo: comboBox categories moved near filter label
Juanjo: comboBox categories can be skined
Juanjo: filter summary (MB and number of files) depends on the selected category 
Juanjo: support for file categories
Juanjo: MsnSkin fixed [roytam1]
Juanjo: Autoscroll download list if the details panel hides the selected item
Juanjo: When the form is resized the details panel size is not increased or shown if it was hidden
Juanjo: show client version info in version.subversion format
Juanjo: Enable/disable file context menu items depending on if the file is started or not
Juanjo: Restore after minimize to tray does not show file details panel
FeuerFrei: added new Hungarian translation [Thanks Decayer-9]
FeuerFrei: fixed duplicated entries in xml [Thanks Decayer-9]
Juanjo: fixed upload speed goes crazy if upload limit is higher than line upload bandwidth
mimonty: Added compression to remote control
mimonty: Added Ip filter to remote control
Juanjo: fixed small remote control issues
Juanjo: start a stopped file when other finalize option. lphant selects highest priority file.
Juanjo: Save window position and size [by fbarbaise]
Juanjo: show a message if no player is configured in preferences
Juanjo: added ClientGTK 1st code [by finrold]
toertchn: added SetRuntimeProperty in Config.cs
toertchn: added SetPath/GetPath in InterfaceGateway
toertchn: StartKernel() added to CKernel & CInterfaceGateway (start in Client\KernelGateway.cs > Initialize)
toertchn: Controls\TaskbarNotifier.cs ~ Line 331 old line in comment > mono don't know Screen.GetWorkingArea(Rect)
toertchn: added namespace eLePhant.Net
toertchn: Firewall.cs in eLePhant.Net (currently not used)
toertchn: nant added to cvs
FeuerFrei: fixed crash with a blank line in ipfilter.dat
FeuerFrei: added ScrollingCredits control
FeuerFrei: added credits in the "About" for developers, retired members and translators
Juanjo: cache version number for better performance
Juanjo: version number is calculated from Protocol instead of Assembly (assembly returns 0.7.0.0 which is not 0.07)
FeuerFrei: new upload slots are opened if needed
toertchn: added CKernel.DoReportBug(string exception);
toertchn: added string CKernel.DllDirectory;
toertchn: added string CKernel.LangDirectory;
toertchn: added string CKernel.Version, string Gateway.Version;
Juanjo: fixed exception processing chunk status
Juanjo: drop elinks into lphant window
FeuerFrei: fixed crash on search with mdf and mds files in the results
FeuerFrei: added total blocked ip message

----------------------
- May 25h, 2004 -
----------------------
Juanjo: solved, a file can be rehashed many times under some contitions.
FeuerFrei: fixed crash when publishing avi file to server
Juanjo: known.met is saved after each rehash instead of after rehashing all files, so if hash is aborted all hashes done are not lost.
FeuerFrei: Blocked IP are now only displayed in the server log in DEBUG mod
FeuerFrei: added a new translation: Finnish [Thanks to Jesse]
FeuerFrei: avoid 0 file size in eLink crash
FeuerFrei: added new translations for B (Byte),KB,MB,GB
FeuerFrei: fixed English translation [Thanks to Jeremy]
FeuerFrei: fixed sorting (download progress and upload) 
toertchn: reorganized some code
Juanjo: MsnSkin changed [roytam1]


----------------------
- May 17h, 2004 -
----------------------
Juanjo: upgraded to udp client <-> client version 4
FeuerFrei: added spam keywords
FeuerFrei: fixed crash with server elink with a high port number
Juanjo: fixed bug #119 Downloading / Hashing a file with the same name. 
FeuerFrei: added a spam filter
FeuerFrei: added a new option in config.xml : "Receive message from friend only"
Juanjo: fixed crash "on invisible control"
FeuerFrei: removed minimize and maximize button in options, tabkernel is selected
FeuerFrei: we can add server with high port number in servers form
Juanjo: fixed bug #92 another exception on show context menu
Juanjo: dont't use udp reask when you have to request sources
Juanjo: change UDP port without restart. 

Source/Client/CSkin.cs
#region Copyright (c)2003 Juanjo < http://lphant.sourceforge.net >
/*
* This file is part of eLePhant
* Copyright (C)2003 Juanjo < j_u_a_n_j_o@users.sourceforge.net / http://lphant.sourceforge.net >
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either
* version 2 of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
#endregion

using System;
using System.Resources;
using System.Drawing;
using System.Windows.Forms;
using System.IO;
using System.Diagnostics;
using System.Xml;

namespace eLePhant.Client
{
	/// <summary>
	/// Summary description for CSkin.
	/// </summary>
	public class CSkin
	{
		private string m_CurrentSkin;
		private XmlElement m_XMLColors;
		private XmlElement m_XMLPositions;
		private XmlElement m_XMLSizes;
		private XmlElement m_XMLDefaultColors;
		private XmlElement m_XMLDefaultPositions;
		private XmlElement m_XMLDefaultSizes;
		private string m_LinkText;
		private string m_LinkUrl;

		public string LinkText
		{
			get
			{
				return m_LinkText;
			}
		}

		public string LinkUrl
		{
			get
			{
				return m_LinkUrl;
			}
		}

		public string CurrentSkin
		{
			get
			{
				return m_CurrentSkin;
			}
			set
			{
				m_CurrentSkin=value;
				if (!m_DefaultSkin()) Load();
				else
				{
					m_LinkText="";
					m_LinkUrl="";
				}
			}
		}
		
		private bool m_DefaultSkin()
		{
			return (m_CurrentSkin=="default");
		}
		
		public CSkin()
		{
			m_CurrentSkin="Default";
			XmlDocument doc = new XmlDocument();
			doc.Load(typeof(eLePhantForm).Assembly.GetManifestResourceStream(typeof(eLePhantForm),"Client.Resources.Classic.skin.xml"));
			m_XMLDefaultColors=doc.DocumentElement["colors"];
			m_XMLDefaultPositions=doc.DocumentElement["positions"];
			m_XMLDefaultSizes=doc.DocumentElement["sizes"];
			m_LinkText="";
			m_LinkUrl="";
		}

		public Bitmap GetBitmap(string imageID)
		{
			bool notFound=false;
			if (!m_DefaultSkin())
			{
				try
				{
					string fileName;
					fileName=Application.StartupPath+Path.DirectorySeparatorChar+"skins\\"+m_CurrentSkin+Path.DirectorySeparatorChar+imageID+".bmp";
					if (File.Exists(fileName))
						return new Bitmap(fileName);
					fileName=Application.StartupPath+Path.DirectorySeparatorChar+"skins\\"+m_CurrentSkin+Path.DirectorySeparatorChar+imageID+".gif";
					if (File.Exists(fileName))
						return new Bitmap(fileName);
					fileName=Application.StartupPath+Path.DirectorySeparatorChar+"skins\\"+m_CurrentSkin+Path.DirectorySeparatorChar+imageID+".ico";
					if (File.Exists(fileName))
						return new Bitmap(fileName);
					fileName=Application.StartupPath+Path.DirectorySeparatorChar+"skins\\"+m_CurrentSkin+Path.DirectorySeparatorChar+imageID+".jpg";
					if (File.Exists(fileName))
						return new Bitmap(fileName);
					else 
						notFound=true;
				}
				catch 
				{
					Debug.WriteLine(imageID+" not found");
				}
			}
			if ((m_DefaultSkin())||(notFound)) 
			{
				string resourceName=imageID+".bmp";
				try
				{
					return new Bitmap(typeof(eLePhantForm),"Client.Resources.Classic."+resourceName);
				}
				catch
				{
					try
					{
						resourceName=imageID+".gif";
						return new Bitmap(typeof(eLePhantForm),"Client.Resources.Classic."+resourceName);
					}
					catch 
					{
						try
						{
							resourceName=imageID+".ico";
							return new Bitmap(typeof(eLePhantForm),"Client.Resources.Classic."+resourceName);
						}
						catch
						{
							try
							{
								resourceName=imageID+".jpg";
								return new Bitmap(typeof(eLePhantForm),"Client.Resources.Classic."+resourceName);
							}
							catch
							{
							}
						}
					}
				}
			}
			return null;
		}

		public Color GetColor(string colorID)
		{
			bool notFound=false;
			if (!m_DefaultSkin())
			{
				try
				{
					XmlNode element=m_XMLColors.SelectSingleNode("color[@key='"+colorID+"']");
					int color=int.Parse(element.Attributes["value"].Value,System.Globalization.NumberStyles.HexNumber);
					return Color.FromArgb(color);
				}
				catch
				{
					Debug.WriteLine(colorID + " not found");
					notFound=true;
				}
			}
			if ((m_DefaultSkin())||(notFound))
			{
				try
				{
					XmlNode element=m_XMLDefaultColors.SelectSingleNode("color[@key='"+colorID+"']");
					int color=int.Parse(element.Attributes["value"].Value,System.Globalization.NumberStyles.HexNumber);
					return Color.FromArgb(color);
				}
				catch
				{
					Debug.WriteLine(colorID + " not found default");
					notFound=true;
				}
			}
			return Color.Gray;
		}

		public void GetPosition(string positionID, ref int x, ref int y)
		{
			bool notFound=false;
			if (!m_DefaultSkin())
			{
				try
				{
					XmlNode element=m_XMLPositions.SelectSingleNode("position[@key='"+positionID+"']");
					x=int.Parse(element.Attributes["x"].Value);
					y=int.Parse(element.Attributes["y"].Value);
					return;
				}
				catch
				{
					Debug.WriteLine(positionID + " not found");
					notFound=true;
				}
			}
			if ((m_DefaultSkin())||(notFound))
			{
				try
				{
					XmlNode element=m_XMLDefaultPositions.SelectSingleNode("position[@key='"+positionID+"']");
					x=int.Parse(element.Attributes["x"].Value);
					y=int.Parse(element.Attributes["y"].Value);
					return;
				}
				catch
				{
					Debug.WriteLine(positionID + " not found default");
					notFound=true;
				}
			}
		}

		public void GetSize(string sizeID, ref int width, ref int heigh)
		{
			bool notFound=false;
			if (!m_DefaultSkin())
			{
				try
				{
					XmlNode element=m_XMLSizes.SelectSingleNode("size[@key='"+sizeID+"']");
					width=int.Parse(element.Attributes["width"].Value);
					heigh=int.Parse(element.Attributes["height"].Value);
					return;
				}
				catch
				{
					Debug.WriteLine(sizeID + " not found");
					notFound=true;
				}
			}
			if ((m_DefaultSkin())||(notFound))
			{
				try
				{
					XmlNode element=m_XMLDefaultSizes.SelectSingleNode("size[@key='"+sizeID+"']");
					width=int.Parse(element.Attributes["width"].Value);
					heigh=int.Parse(element.Attributes["height"].Value);
					return;
				}
				catch
				{
					Debug.WriteLine(sizeID + " not found default");
					notFound=true;
				}
			}
		}

		public bool Load()
		{
			string version;
			try
			{
				XmlDocument doc = new XmlDocument();
				doc.Load(Application.StartupPath+Path.DirectorySeparatorChar+"skins\\"+m_CurrentSkin+Path.DirectorySeparatorChar + "skin.xml");
				
				version = doc.DocumentElement.Attributes["version"].InnerText;
				Debug.WriteLine("Version: "+version);
				m_XMLColors=doc.DocumentElement["colors"];
				m_XMLSizes=doc.DocumentElement["sizes"];
				m_XMLPositions=doc.DocumentElement["positions"];
				XmlElement xmlURL=doc.DocumentElement["URL"];
				m_LinkText="";
				m_LinkUrl="";
				if (xmlURL!=null)
				{
					XmlNode nodeURL=xmlURL.FirstChild;
					if (nodeURL!=null)
					{
						this.m_LinkText=nodeURL.Attributes["Text"].Value;
						this.m_LinkUrl=nodeURL.Attributes["url"].Value;
					}
				}

				return true;
			}
			catch (Exception e)
			{
				Debug.WriteLine("Exception while load Skin:\n " + e.ToString());
				m_CurrentSkin="default";
				return false;
			}			
		}
	}
}
